pkg/driver/xovis/udmi.go

```diff
@@ -0,0 +1,150 @@
+	Version       string      `json:"version"`
+	Timestamp     time.Time   `json:"timestamp"`
+	PartialUpdate bool        `json:"partial_update,omitempty"`
+	Points        EventPoints `json:"points"`
+}
+
+type EventPoints struct {
+	// Generic points
+	DeviceType *EventPoint[string] `json:"DeviceType,omitempty"`
+
+	// footfall count related points
+	EnterCount *EventPoint[int32] `json:"EnterCount,omitempty"`
+	LeaveCount *EventPoint[int32] `json:"LeaveCount,omitempty"`
+
+	// occupancy related points
+	PeopleCount    *EventPoint[int32]  `json:"PeopleCount,omitempty"`
+	OccupancyState *EventPoint[string] `json:"OccupancyState,omitempty"`
+}
+
+type EventPoint[T any] struct {
+	PresentValue T `json:"present_value"`
+}
+type UdmiServiceServer struct {
+	gen.UnimplementedUdmiServiceServer
+
+	logger *zap.Logger
+
+	enterLeave      *resource.Value
+	occupancy       *resource.Value
+	udmiTopicPrefix string
+}
+
+func NewUdmiServiceServer(logger *zap.Logger, e *resource.Value, o *resource.Value, udmiPrefix string) *UdmiServiceServer {
+	return &UdmiServiceServer{
+		logger:          logger,
+		enterLeave:      e,
+		occupancy:       o,
+		udmiTopicPrefix: udmiPrefix,
+	}
+}
+
+func (u *UdmiServiceServer) PullControlTopics(_ *gen.PullControlTopicsRequest, _ gen.UdmiService_PullControlTopicsServer) error {
+	// we don't have any control topics
+	return status.Error(codes.Unimplemented, "not implemented")
+}
+func (u *UdmiServiceServer) OnMessage(_ context.Context, _ *gen.OnMessageRequest) (*gen.OnMessageResponse, error) {
+	// we don't support doing anything here
+	return nil, status.Error(codes.Unimplemented, "not implemented")
+}
+
+func (u *UdmiServiceServer) GetExportMessage(_ context.Context, _ *gen.GetExportMessageRequest) (*gen.MqttMessage, error) {
+	return nil, status.Error(codes.Unimplemented, "not implemented")
+}
+
+func (u *UdmiServiceServer) PullExportMessages(request *gen.PullExportMessagesRequest, server gen.UdmiService_PullExportMessagesServer) error {
+	ctx, cancel := context.WithCancel(server.Context())
+	defer cancel()
+
+	enterLeaveChanges := make(<-chan *resource.ValueChange)
+	if u.enterLeave != nil {
+		enterLeaveChanges = u.enterLeave.Pull(ctx,
+			resource.WithUpdatesOnly(true),
+		)
+	}
+
+	occupancyChanges := make(<-chan *resource.ValueChange)
+	if u.occupancy != nil {
+		occupancyChanges = u.occupancy.Pull(ctx,
+			resource.WithUpdatesOnly(true),
+		)
+	}
+
+	for {
+		var eventPoints = EventPoints{
+			DeviceType: &EventPoint[string]{PresentValue: DriverName},
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		case change, _ := <-enterLeaveChanges:
+			if change != nil {
+				airQuality := change.Value.(*traits.EnterLeaveEvent)
+				appendEnterLeaveEventPoints(airQuality, &eventPoints)
+			}
+
+		case change, _ := <-occupancyChanges:
+			if change != nil {
+				temperature := change.Value.(*traits.Occupancy)
+				appendOccupancyEventPoints(temperature, &eventPoints)
+			}
+		}
```

Comment:
When a channel closes, selects on it will complete immediately, which can cause a busy loop. It's not a big problem here because these channels only close when the context is cancelled, which would be caught by the top case shortly afterwards. But in general, when you have channels that might close, you should do something like this to avoid the busy-loop problem:
```go
for {
    select {
    case x, ok := <-ch1:
        if !ok {
            ch1 = nil
            break
        }
        process(x)
    case x, ok := <-ch2:
        if !ok {
            ch2 = nil
            break
        }
        process(x)
}
```

If you decide it's not a problem, then prefer `change := <-occupancyChanges` to `change, _ := <-occupancyChanges` - they have the same meaning.